cql3/restrictions/statement_restrictions.cc

```diff
@@ -598,14 +600,93 @@ void statement_restrictions::process_clustering_columns_restrictions(bool for_vi
     }
 }
 
+namespace {
+
+using namespace expr;
+
+/// Computes partition-key ranges from token atoms in ex.
+dht::partition_range_vector partition_ranges_from_token(const expr::expression& ex, const query_options& options) {
+    auto values = possible_lhs_values(nullptr, ex, options);
+    if (values == expr::value_set(expr::value_list{})) {
+        return {};
+    }
+    const auto bounds = expr::to_range(values);
+    const auto start_token = bounds.start() ? dht::token::from_bytes(bounds.start()->value())
+            : dht::minimum_token();
+    auto end_token = bounds.end() ? dht::token::from_bytes(bounds.end()->value()) : dht::maximum_token();
+    const bool include_start = bounds.start() && bounds.start()->is_inclusive();
+    const auto include_end = bounds.end() && bounds.end()->is_inclusive();
+
+    auto start = dht::partition_range::bound(include_start
+                       ? dht::ring_position::starting_at(start_token)
+                       : dht::ring_position::ending_at(start_token));
+    auto end = dht::partition_range::bound(include_end
+                     ? dht::ring_position::ending_at(end_token)
+                     : dht::ring_position::starting_at(end_token));
+
+    return {{std::move(start), std::move(end)}};
+}
+
+/// Turns a partition-key value into a partition_range. \p pk must have elements for all partition columns.
+dht::partition_range range_from_bytes(const schema& schema, const std::vector<bytes>& pk) {
+    const auto k = partition_key::from_exploded(pk);
+    const query::ring_position pos(dht::get_token(schema, k), k);
```

Comment:
std::move(k) (which means you'll need to get_token earlier).

Reply:
Done.